lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php

```diff
@@ -0,0 +1,832 @@
+		if ( !$this->canGroupBeRemoved( $userPerformingChange, $userToChange, $group ) ) {
+			return false;
+		}
+
+		if ( in_array( $group, $this->getGlobalGroups() ) ) {
+			return $this->removeUserFromGlobalGroup( $userToChange, $group );
+		} else {
+			return $this->removeUserFromLocalGroup( $userToChange, $group );
+		}
+	}
+
+	private function invalidateUserGroupsAndPermissions( $userId ) {
+		unset( $this->userPermissions[$userId] );
+		unset( $this->localExplicitUserGroups[$userId] );
+		unset( $this->implicitUserGroups[$userId] );
+		unset( $this->globalExplicitUserGroups[$userId] );
+	}
+
+
+	public function doesUserHavePermission( \User $user, $permission ) {
+		if ( $permission === '' ) {
+			return true; // In the spirit of DWIM
+		}
+		# Patrolling may not be enabled
+		if( $permission === 'patrol' || $permission === 'autopatrol' ) {
+			global $wgUseRCPatrol, $wgUseNPPatrol;
+			if( !$wgUseRCPatrol && !$wgUseNPPatrol )
+				return false;
+		}
+		# Use strict parameter to avoid matching numeric 0 accidentally inserted
+		# by misconfiguration: 0 == 'foo'
+		return in_array( $permission, $this->getUserPermissions( $user ), true );
+	}
+
+	public function doesUserHaveAllPermissions( \User $user, $permissions ) {
+		foreach( $permissions as $permission ){
+			if( !$this->doesUserHavePermission( $user, $permission ) ){
+				return false;
+			}
+		}
+		return true;
+	}
+
+	public function doesUserHaveAnyPermission( \User $user, $permissions ) {
+		foreach( $permissions as $permission ){
+			if( $this->doesUserHavePermission( $user, $permission ) ){
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Returns an array of groups that this user can add and remove
+	 * @param \User user
+	 * @return Array array( 'add' => array( addablegroups ),
+	 *  'remove' => array( removablegroups ),
+	 *  'add-self' => array( addablegroups to self),
+	 *  'remove-self' => array( removable groups from self) )
+	 */
+	public function getGroupsChangeableByUser( \User $user ) {
+		if( $this->doesUserHavePermission( $user, 'userrights' ) ) {
+			// This group gives the right to modify everything (reverse-
+			// compatibility with old "userrights lets you change
+			// everything")
+			// Using array_merge to make the groups reindexed
+			$all = array_merge( $this->getExplicitGroups() );
+			return array(
+				'add' => $all,
+				'remove' => $all,
+				'add-self' => array(),
+				'remove-self' => array()
+			);
+		}
+
+		// Okay, it's not so simple, we will have to go through the arrays
+		$groups = array(
+			'add' => array(),
+			'remove' => array(),
+			'add-self' => array(),
+			'remove-self' => array()
+		);
+		$addergroups = $this->getEffectiveUserGroups( $user );
+
+		foreach( $addergroups as $addergroup ) {
+			$groups = array_merge_recursive(
+				$groups, $this->getGroupsChangeableByGroup( $addergroup )
+			);
+			$groups['add']    = array_unique( $groups['add'] );
```

Comment:
I guess we could do a single pass of `array_unique` calls after the `foreach` loop, is that correct?